timemachine/fe/mcgregor.py

```diff
@@ -244,13 +167,183 @@ def to_networkx(self) -> nx.Graph:
+
+        if new_v2 == UNMAPPED:
+            # Set to False rows corresponding to the edges of new_v1
+            new_marcs[e1] = False
+        else:
+            e2 = g2.get_edges_as_vector(new_v2)
+
+            # Set to False edge pairs (e1, e2) where either
+            # 1. e1 connects to new_v1 but e2 does NOT connect to new_v2
+            # 2. e1 does NOT connect to new_v1 but e2 connects to new_v2
+            mask = e1[:, None] == e2[None, :]
+            new_marcs &= mask
+
+        return Marcs(new_marcs)
+
+
+@dataclass(frozen=True)
+class AtomMap:
+    a_to_b: Tuple[int, ...]
+    b_to_a: Tuple[int, ...]
+
+    @classmethod
+    def init(cls, n_1: int, n_2: int) -> "AtomMap":
+        return cls((UNMAPPED,) * n_1, (UNMAPPED,) * n_2)
+
+    def add(self, new_v1: int, new_v2: int) -> "AtomMap":
+        def set_at(xs: Tuple[int, ...], idx: int, val: int) -> Tuple[int, ...]:
+            return xs[:idx] + (val,) + xs[idx + 1 :]
+
+        return AtomMap(
+            set_at(self.a_to_b, new_v1, new_v2),
+            set_at(self.b_to_a, new_v2, new_v1),
+        )
+
+    @cached_property
+    def core_size(self):
+        return sum(1 for j in self.a_to_b if j != UNMAPPED)
+
+
+def _verify_map_preserves_core_edges(g1: Graph, g2: Graph, new_v1: int, new_v2: int, atom_map: AtomMap) -> bool:
+    def verify(g1: Graph, g2: Graph, new_v1: int, map_1_to_2: Sequence[int]):
+        for e1 in g1.get_edges(new_v1):
+            src, dst = g1.edges[e1]
+            src_2, dst_2 = map_1_to_2[src], map_1_to_2[dst]
+            if src_2 != UNMAPPED and dst_2 != UNMAPPED:
+                # both ends are mapped
+                # see if this edge is present in g2
+                if not g2.cmat[src_2][dst_2]:
+                    return False
+        return True
+
+    return verify(g1, g2, new_v1, atom_map.a_to_b) and verify(g2, g1, new_v2, atom_map.b_to_a)
+
+
+@dataclass(frozen=True)
+class Node:
+    atom_map: AtomMap
+    marcs: Marcs
+    layer: int
+
+    @classmethod
+    def init(cls, g1: Graph, g2: Graph, predicate: NDArray[np.bool_]) -> "Node":
+        return cls(AtomMap.init(g1.n_vertices, g2.n_vertices), Marcs.from_predicate(g1, g2, predicate), 0)
+
+    def add(self, g1: Graph, g2: Graph, new_v2: int) -> "Node":
+        return Node(
+            self.atom_map.add(self.layer, new_v2),
+            self.marcs.refine(g1, g2, self.layer, new_v2),
+            self.layer + 1,
+        )
+
+    def skip(self, g1: Graph, g2: Graph) -> "Node":
+        return Node(
+            self.atom_map,
+            self.marcs.refine(g1, g2, self.layer, UNMAPPED),
+            self.layer + 1,
+        )
+
+    @cached_property
+    def is_leaf(self):
+        return self.layer == len(self.atom_map.a_to_b)
+
+    @cached_property
+    def priority(self):
+        """Compute the priority of this node. By convention, lowest numerical value is highest priority"""
+        return (-self.marcs.num_edges_upper_bound, -self.layer)
```

Comment:
Interesting - I wonder if we can tie-break the equal priority further, eg:

```
(-self.marcs.num_edges_upper_bound, -self.marcs.num_of_ones, +self.marcs.num_connected_components, -self.layer)
```

Reply:
Interesting idea to incorporate the number of connected components. I want to play around with this, but will leave for a future PR as I think this will require additional bookkeeping to be fast enough.